src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockUtils.java

```diff
@@ -0,0 +1,141 @@
+     */
+    @ApiStatus.Internal
+    public static boolean isCodeBlockStartChar(@NotNull PsiFile file, @Nullable Character character) {
+        return (character != null) && getBracePairsFwd(file).containsKey(character);
+    }
+
+    /**
+     * Determines whether or not the specified character can end a code block in the provided file.
+     *
+     * @param file      the PSI file
+     * @param character the optional character
+     * @return true if the character can end a code block in the file; otherwise false
+     */
+    @ApiStatus.Internal
+    public static boolean isCodeBlockEndChar(@NotNull PsiFile file, @Nullable Character character) {
+        return (character != null) && getBracePairsBwd(file).containsKey(character);
+    }
+
+    /**
+     * Returns the code block start character for the specified code block end character in the provided file.
+     *
+     * @param file             the PSI file
+     * @param codeBlockEndChar the optional code block end character
+     * @return the corresponding code block start character if found; otherwise null
+     */
+    @Nullable
+    @ApiStatus.Internal
+    public static Character getCodeBlockStartChar(@NotNull PsiFile file, @Nullable Character codeBlockEndChar) {
+        return codeBlockEndChar != null ? getBracePairsBwd(file).get(codeBlockEndChar) : null;
+    }
+
+    /**
+     * Returns the code block end character for the specified code block start character in the provided file.
+     *
+     * @param file               the PSI file
+     * @param codeBlockStartChar the optional code block start character
+     * @return the corresponding code block end character if found; otherwise null
+     */
+    @Nullable
+    @ApiStatus.Internal
+    public static Character getCodeBlockEndChar(@NotNull PsiFile file, @Nullable Character codeBlockStartChar) {
+        return codeBlockStartChar != null ? getBracePairsFwd(file).get(codeBlockStartChar) : null;
+    }
+
+    @NotNull
+    private static Map<Character, Character> getBracePairsFwd(@NotNull PsiFile file) {
+        return CachedValuesManager.getCachedValue(file, () -> {
+            Map<Character, Character> bracePairs = file instanceof TextMateFile ? LSPIJTextMateUtils.getBracePairs(file) :
+                    file.getFileType() instanceof AbstractFileType ? getAbstractFileTypeBracePairs(file) :
+                            null;
+            if (bracePairs == null) {
+                bracePairs = DEFAULT_BRACE_PAIRS;
+            }
+            return Result.create(bracePairs, file);
+        });
+    }
+
+    @NotNull
+    private static Map<Character, Character> getBracePairsBwd(@NotNull PsiFile file) {
+        return CachedValuesManager.getCachedValue(file, () -> Result.create(ContainerUtil.reverseMap(getBracePairsFwd(file)), file));
+    }
+
+    @Nullable
+    private static Map<Character, Character> getAbstractFileTypeBracePairs(@NotNull PsiFile file) {
```

Comment:
This is how supported brace pairs are derived for plain text files backed by `AbstractFileType`s.